docs/fundamentals/code-analysis/quality-rules/ca2020.md

```diff
@@ -0,0 +1,159 @@
+ms.date: 10/07/2022
+ms.topic: reference
+f1_keywords:
+  - PreventNumericIntPtrUIntPtrBehavioralChanges 
+  - CA2020
+helpviewer_keywords:
+  - PreventNumericIntPtrUIntPtrBehavioralChanges 
+  - CA2020
+author: buyaa-n
+---
+# CA2020: Prevent behavioral change caused by built-in operators of IntPtr/UIntPtr
+
+| | Value |
+|-|-|
+| **Rule ID** |CA2020|
+| **Category** |[Reliability](reliability-warnings.md)|
+| **Fix is breaking or non-breaking** |Non-breaking|
+
+## Cause
+
+This rule fires when it detects a behavioral change between .NET 6.0 and .NET 7.0 introduced with new built-in operators of numeric IntPtr/UIntPtr feature.
+
+## Rule description
+
+With the [numeric IntPtr feature](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-11.0/numeric-intptr.md), System.IntPtr and System.UIntPtr gained some built-in operators (conversions, unary and binary). Those now might throw when overflowing within checked context or may not throw in unchecked context compared to the previous `user-defined` operators in .NET 6.0 and below, which could cause behavioral change when upgrading from .NET 6.0 and below into .NET 7.
+
+### List of APIs affected
+
+| Operator | Context | In .NET 7 | In .NET 6 and earlier | Example |
+|---|---|---|---|---|
+| operator +(IntPtr, int) | Checked | Throws when overflows | Doesn't throw when overflows | `checked(intPtrVariable + 2);` |
+| operator -(IntPtr, int) | Checked | Throws when overflows | Doesn't throw when overflows | `checked(intPtrVariable - 2);` |
+| explicit operator IntPtr(long) | Unchecked | Doesn't throw when overflows | Can throw in 32-bit contexts | `(IntPtr)longVariable;` |
+| explicit operator void\*(IntPtr) | Checked | throws when overflows | Doesn't throw when overflows | `checked((void*)intPtrVariable);` |
+| explicit operator IntPtr(void\*) | Checked | throws when overflows | Doesn't throw when overflows | `checked((IntPtr)voidPtrVariable);` |
+| explicit operator int(IntPtr) | Unchecked | Doesn't throw when overflows | Can throw in 64-bit contexts | `(int)intPtrVariable;` |
+| operator +(UIntPtr, int) | Checked | Throws when overflows | Doesn't throw when overflows | `checked(uintPtrVariable + 2);` |
+| operator -(UIntPtr, int) | Checked | Throws when overflows | Doesn't throw when overflows | `checked(uintPtrVariable - 2);` |
+| explicit operator UIntPtr(ulong) | Unchecked | Doesn't throw when overflows | Can throw in 32-bit contexts | `(UIntPtr)uLongVariable` |
+| explicit operator uint(UIntPtr) | Unchecked | Doesn't throw when overflows | Can throw in 64-bit contexts | `(uint)uintPtrVariable` |
```

Comment:
I think Checked/Unchecked better lower cased same as the checked/unchecked keywords

```suggestion
| operator +(IntPtr, int) | checked | Throws when overflows | Doesn't throw when overflows | `checked(intPtrVariable + 2);` |
| operator -(IntPtr, int) | checked | Throws when overflows | Doesn't throw when overflows | `checked(intPtrVariable - 2);` |
| explicit operator IntPtr(long) | unchecked | Doesn't throw when overflows | Can throw in 32-bit contexts | `(IntPtr)longVariable;` |
| explicit operator void\*(IntPtr) | checked | throws when overflows | Doesn't throw when overflows | `checked((void*)intPtrVariable);` |
| explicit operator IntPtr(void\*) | checked | throws when overflows | Doesn't throw when overflows | `checked((IntPtr)voidPtrVariable);` |
| explicit operator int(IntPtr) | unchecked | Doesn't throw when overflows | Can throw in 64-bit contexts | `(int)intPtrVariable;` |
| operator +(UIntPtr, int) | checked | Throws when overflows | Doesn't throw when overflows | `checked(uintPtrVariable + 2);` |
| operator -(UIntPtr, int) | checked | Throws when overflows | Doesn't throw when overflows | `checked(uintPtrVariable - 2);` |
| explicit operator UIntPtr(ulong) | unchecked | Doesn't throw when overflows | Can throw in 32-bit contexts | `(UIntPtr)uLongVariable` |
| explicit operator uint(UIntPtr) | unchecked | Doesn't throw when overflows | Can throw in 64-bit contexts | `(uint)uintPtrVariable` |
```